execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -1067,25 +1070,39 @@ where
             mut path,
         }) = self.visited.pop()
         {
-            let mut maybe_next_trie: Option<Trie<K, V>> = None;
+            let mut maybe_next_trie: Option<LazyTrieLeaf<K, V>> = None;
 
             match trie {
-                Trie::Leaf { key, .. } => {
-                    let key_bytes = match key.to_bytes() {
-                        Ok(bytes) => bytes,
-                        Err(e) => {
-                            self.state = KeysIteratorState::Failed;
-                            return Some(Err(e.into()));
-                        }
-                    };
-                    debug_assert!(key_bytes.starts_with(&path));
+                LazyTrieLeaf::Left(leaf_bytes) => {
+                    if leaf_bytes.is_empty() {
+                        self.state = KeysIteratorState::Failed;
+                        return Some(Err(bytesrepr::Error::Formatting.into()));
+                    }
+
+                    let key_bytes = &leaf_bytes[1..]; // Skip `Trie::Leaf` tag
+                    debug_assert!(
+                        key_bytes.starts_with(&path),
+                        "Expected key bytes to start with the current path"
+                    );
+
                     // only return the leaf if it matches the initial descend path
                     path.extend(&self.initial_descend);
                     if key_bytes.starts_with(&path) {
+                        // Only deserializes K when we're absolutely sure the path matches.
+                        let (key, _stored_value): (K, _) = match K::from_bytes(key_bytes) {
+                            Ok(key) => key,
+                            Err(error) => {
+                                self.state = KeysIteratorState::Failed;
+                                return Some(Err(error.into()));
+                            }
+                        };
                         return Some(Ok(key));
                     }
                 }
-                Trie::Node { ref pointer_block } => {
+                LazyTrieLeaf::Right(Trie::Leaf { .. }) => {
+                    unreachable!("Lazy trie deserializer ensures that this variant never happens.")
```

Comment:
I'm wondering if we can encode this in the type system, so that this `unreachable!` arm is no longer needed.

Reply:
We could, something like `enum LazyTrie<K, V> { Leaf(Bytes, PhantomData<(K, V)>), Node(TrieNode), Extension(TrieExtension) }` etc. Now that we'll do more lazy trie leaf deserialization I added a ticket to deal with this

Reply:
Ticket https://github.com/casper-network/casper-node/issues/3994